File: EPRO.Library.v3.5/EPRO.Library.v3.5/Modules/basObjectNarrator.cs
using System;
using System.Linq.Expressions;

namespace ELibrary.Standard.Modules
{
    public static class basObjectNarrator
    {

        /// <summary>
        /// Usage: VB: Function() variableName, C# ()=> variableName
        /// </summary>
        /// <typeparam name="T"></typeparam>
        /// <param name="expr"></param>
        /// <returns></returns>
        /// <remarks></remarks>
        public static string GetVariableFullName<T>(Expression<Func<T>> expr)
        {
            MemberExpression body = (MemberExpression)expr.Body;
            if (body.ToString().IndexOf("+") >= 0)
                return body.Member.Name; // REM it is a local variable
            return body.ToString().Split(new string[] { ", " }, StringSplitOptions.RemoveEmptyEntries)[1].Split(new string[] { ": " }, StringSplitOptions.RemoveEmptyEntries)[1];
        }

        /// <summary>
        /// Usage: VB: Function() variableName, C# ()=> variableName
        /// </summary>
        /// <typeparam name="T"></typeparam>
        /// <param name="expr"></param>
        /// <returns></returns>
        /// <remarks></remarks>
        public static string GetVariableName<T>(Expression<Func<T>> expr)
        {
            MemberExpression body = (MemberExpression)expr.Body;
            return body.Member.Name;
        }

        /// <summary>
        /// Usage: Function() Me.MethodName(). if there is param, pass null
        /// </summary>
        /// <typeparam name="T"></typeparam>
        /// <param name="expr"></param>
        /// <returns></returns>
        /// <remarks></remarks>
        public static string GetMethodFullName<T>(Expression<Func<T>> expr)
        {
            MethodCallExpression body = (MethodCallExpression)expr.Body;
            return body.ToString().Split(new string[] { ", " }, StringSplitOptions.RemoveEmptyEntries)[1].Split(new string[] { ": " }, StringSplitOptions.RemoveEmptyEntries)[1];
        }

        /// <summary>
        /// Usage: Function() Me.MethodName(). if there is param, pass null
        /// </summary>
        /// <typeparam name="T"></typeparam>
        /// <param name="expr"></param>
        /// <returns></returns>
        /// <remarks></remarks>
        public static string GetMethodName<T>(Expression<Func<T>> expr)
        {
            MethodCallExpression body = (MethodCallExpression)expr.Body;
            return body.Method.Name;
        }

        public static string GetTypeFullName<T>(T obj)
        {
            return obj.GetType().FullName;
        }

        public static string GetTypeName<T>(T obj)
        {
            return obj.GetType().Name;
        }
    }
}

File: EPRO.Library.v3.5/EPRO.Library.v3.5/Modules/basObjectNarrator.vb
Imports System.Linq.Expressions

Namespace Modules

    Public Module basObjectNarrator

        ''' <summary>
        ''' Usage: VB: Function() variableName, C# ()=> variableName
        ''' </summary>
        ''' <typeparam name="T"></typeparam>
        ''' <param name="expr"></param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Function GetVariableFullName(Of T)(ByVal expr As Expression(Of Func(Of T))) As String
            Dim body As MemberExpression = CType(expr.Body, MemberExpression)
            If body.ToString().IndexOf("+") >= 0 Then Return body.Member.Name REM it is a local variable
            Return body.ToString().Split(New String() {", "}, StringSplitOptions.RemoveEmptyEntries)(1).Split(New String() {": "}, StringSplitOptions.RemoveEmptyEntries)(1)
        End Function

        ''' <summary>
        ''' Usage: VB: Function() variableName, C# ()=> variableName
        ''' </summary>
        ''' <typeparam name="T"></typeparam>
        ''' <param name="expr"></param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Function GetVariableName(Of T)(ByVal expr As Expression(Of Func(Of T))) As String
            Dim body As MemberExpression = CType(expr.Body, MemberExpression)
            Return body.Member.Name
        End Function

        ''' <summary>
        ''' Usage: Function() Me.MethodName(). if there is param, pass null
        ''' </summary>
        ''' <typeparam name="T"></typeparam>
        ''' <param name="expr"></param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Function GetMethodFullName(Of T)(ByVal expr As Expression(Of Func(Of T))) As String
            Dim body As MethodCallExpression = CType(expr.Body, MethodCallExpression)
            Return body.ToString().Split(New String() {", "}, StringSplitOptions.RemoveEmptyEntries)(1).Split(New String() {": "}, StringSplitOptions.RemoveEmptyEntries)(1)
        End Function

        ''' <summary>
        ''' Usage: Function() Me.MethodName(). if there is param, pass null
        ''' </summary>
        ''' <typeparam name="T"></typeparam>
        ''' <param name="expr"></param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Function GetMethodName(Of T)(ByVal expr As Expression(Of Func(Of T))) As String
            Dim body As MethodCallExpression = CType(expr.Body, MethodCallExpression)
            Return body.Method.Name
        End Function

        Public Function GetTypeFullName(Of T)(ByVal obj As T) As String
            Return obj.GetType().FullName
        End Function

        Public Function GetTypeName(Of T)(ByVal obj As T) As String
            Return obj.GetType().Name
        End Function



    End Module

End Namespace